File: android/app/src/main/java/com/mistroom/core/mesh/DedupCache.kt
package com.mistroom.core.mesh

import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Deduplication cache for mesh packets.
 *
 * Prevents the same packet from being processed or forwarded multiple times
 * as it propagates through the mesh network. Uses packet IDs with TTL-based
 * expiration.
 */
@Singleton
class DedupCache @Inject constructor() {

    companion object {
        /** Default entry expiration time: 5 minutes. */
        const val DEFAULT_EXPIRY_MILLIS = 5 * 60 * 1000L

        /** Maximum cache size before forced cleanup. */
        const val MAX_CACHE_SIZE = 10_000
    }

    private data class CacheEntry(
        val packetId: String,
        val insertedAtMillis: Long = System.currentTimeMillis(),
    )

    private val cache = ConcurrentHashMap<String, CacheEntry>()
    private var expiryMillis = DEFAULT_EXPIRY_MILLIS

    /**
     * Check if a packet ID has been seen before.
     * If not seen, it is added to the cache atomically.
     *
     * @param packetId The unique packet identifier.
     * @return `true` if this packet ID was already in the cache (duplicate),
     *         `false` if it's new (first time seen — now added).
     */
    fun isDuplicate(packetId: String): Boolean {
        evictExpired()

        val entry = CacheEntry(packetId)
        val previous = cache.putIfAbsent(packetId, entry)
        return previous != null
    }

    /**
     * Explicitly mark a packet ID as seen.
     */
    fun markSeen(packetId: String) {
        cache[packetId] = CacheEntry(packetId)
        if (cache.size > MAX_CACHE_SIZE) {
            evictExpired()
        }
    }

    /**
     * Check if a packet ID is in the cache without modifying it.
     */
    fun contains(packetId: String): Boolean {
        evictExpired()
        return cache.containsKey(packetId)
    }

    /**
     * Remove expired entries from the cache.
     */
    fun evictExpired() {
        val cutoff = System.currentTimeMillis() - expiryMillis
        cache.entries.removeIf { it.value.insertedAtMillis < cutoff }
    }

    /**
     * Clear the entire cache.
     */
    fun clear() {
        cache.clear()
    }

    /**
     * Current number of entries in the cache.
     */
    val size: Int get() = cache.size

    /**
     * Set custom expiry duration.
     */
    fun setExpiryMillis(millis: Long) {
        expiryMillis = millis
    }
}

File: web-client/src/core/mesh/DedupCache.ts
// DedupCache.ts
// Prevents reprocessing or re-forwarding duplicate packets in the mesh network.
// Uses a bounded LRU-style set keyed by packetId (hex SHA-256 string).

const DEFAULT_MAX_SIZE = 1024;
const DEFAULT_TTL_MS = 5 * 60 * 1000; // 5 minutes

interface CacheEntry {
  expiresAt: number;
}

export class DedupCache {
  private readonly cache = new Map<string, CacheEntry>();
  private readonly maxSize: number;
  private readonly ttlMs: number;

  constructor(maxSize = DEFAULT_MAX_SIZE, ttlMs = DEFAULT_TTL_MS) {
    this.maxSize = maxSize;
    this.ttlMs = ttlMs;
  }

  /** Returns true if this packetId has already been seen (and not expired) */
  has(packetId: string): boolean {
    const entry = this.cache.get(packetId);
    if (!entry) return false;
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(packetId);
      return false;
    }
    return true;
  }

  /** Mark a packetId as seen */
  add(packetId: string): void {
    // Evict oldest entry if at capacity
    if (this.cache.size >= this.maxSize) {
      const oldest = this.cache.keys().next().value;
      if (oldest) this.cache.delete(oldest);
    }
    this.cache.set(packetId, { expiresAt: Date.now() + this.ttlMs });
  }

  /** Attempt to add; returns false if it was already seen (duplicate) */
  tryAdd(packetId: string): boolean {
    if (this.has(packetId)) return false;
    this.add(packetId);
    return true;
  }

  get size(): number {
    return this.cache.size;
  }

  clear(): void {
    this.cache.clear();
  }
}
